host/volume/zfs/zfs.go

```diff
@@ -452,11 +457,8 @@ func (b *Provider) RestoreVolumeState(volInfo *volume.Info, data json.RawMessage
 		dataset:   dataset,
 		basemount: record.Basemount,
 	}
-	// zfs should have already remounted filesystems; special remount case for snapshots
-	if v.IsSnapshot() {
-		if err := b.mountSnapshot(v); err != nil {
-			return nil, err
-		}
+	if err := b.mountDataset(v); err != nil {
```

Comment:
Just realised it doesn't matter if it's idempotent or not, `mountDataset()` does a check using `stat()` to check the device/fsid of the parent directory to determine if they 2 are different filesystems to determine if it's already mounted.